### ISSUE TITLE
[QUESTION] attention_mask is None in middle pipeline stage?

### ISSUE THREAD

**Issue by XLzed** [NVIDIA/Megatron-LM#752]
The attention_mask is None in middle pipeline stage when the PP_size>=4，and the p2p communication send/recv hidden_states only, which made the results are different between different parallelism configs. Is there any workaround to solve it?
```
def get_batch(data_iterator):
    """Generate a batch."""

    # TODO: this is pretty hacky, find a better way
    if (not mpu.is_pipeline_first_stage()) and (not mpu.is_pipeline_last_stage()):
        return None, None, None, None, None

    # get batches based on the TP rank you are on
    batch = get_batch_on_this_tp_rank(data_iterator)

    # slice batch along sequence dimension for context parallelism
    batch = get_batch_on_this_cp_rank(batch)

    return batch.values()
```

**Comment by ethanhe42** [comment]
if you use mcore model, the attention mask is generated by transformer engine

**Comment by XLzed** [comment]
> if you use mcore model, the attention mask is generated by transformer engine

Thanks! The results are same when enable transformer engine, maybe there need a cached attention_mask in pretrain_gpt.py like nemo.

**Comment by Pikachu1412** [comment]
Why the last rank need get_batch_on_this_tp_rank?  Doesn't this fetch another batch?
